src/Sarif.Multitool/AbsoluteUriCommand.cs

```diff
@@ -55,7 +55,7 @@ public int Run(AbsoluteUriOptions absoluteUriOptions)
         private IEnumerable<AbsoluteUriFile> GetSarifFiles(AbsoluteUriOptions absoluteUriOptions)
         {
             // Get files names first, as we may write more sarif logs to the same directory as we rebase them.
-            HashSet<string> fileNames = CreateTargetsSet(absoluteUriOptions.TargetFileSpecifiers, absoluteUriOptions.Recurse);
+            HashSet<string> fileNames = CreateTargetsSet(absoluteUriOptions.TargetFileSpecifiers, absoluteUriOptions.Recurse, _fileSystem);
```

Comment:
>_fileSystem [](start = 126, length = 11)

It's such a pity that every derived command class accepts, and stores separately, a `FileSystem` object, and then has to pass it down to a bunch of `static` methods on the base class.

What we should have done was (using `RebaseUriCommand` as an example):
```
internal class RebaseUriCommand : CommandBase
{
    public RebaseUriCommand(IFileSystem fileSystem = null) : base(fileSystem)
    {
        ...
    }

    public int Run(RebaseUriOptions rebaseOptions)
    {
        ...
        //  NO NEED TO PASS _fileSystem DOWN TO THE BASE CLASS.
        WriteSarifFile(sarifLog.Log, outputName, formatting);
    }

    ...
}

public abstract class CommandBase
{
    // STORED ONCE, IN BASE CLASS; AVAILABLE TO DERIVED CLASSES IF NEEDED
    protected IFileSystem FileSystem { get; }

    public CommandBase(IFileSystem fileSystem)
    {
        FileSystem = fileSystem;
    }

    // NO LONGER TAKES AN IFileSystem ARGUMENT; NO LONGER STATIC.
    public T ReadSarifFile<T>(string filePath, IContractResolver contractResolver = null)
    {
        var serializer = new JsonSerializer() { ContractResolver = contractResolver };

        // USES THE NEW FileSystem PROPERTY.
        using (JsonTextReader reader = new JsonTextReader(new StreamReader(FileSystem.OpenRead(filePath))))
        {
            return serializer.Deserialize<T>(reader);
        }
    }
    ...
}
```

Of course it would be a breaking change to remove the existing public static methods on the base class. But we could keep the static methods, _add_ new same-named instance methods, and have the instance methods delegate to the static methods.

I don't recommend we do this now. Now that I've thought it through, I'll file an issue.

Reply:
I filed [https://github.com/microsoft/sarif-sdk/issues/1550](https://github.com/microsoft/sarif-sdk/issues/1550), "Store IFileSystem object in protected property of CommandBase".

---
In reply to: [298665212](https://github.com/microsoft/sarif-sdk/pull/1549#discussion_r298665212) [](ancestors = 298665212)